libs/auth/src/inserted_smart_card_auth.test.ts

```diff
@@ -1,20 +1,928 @@
+      expectedAuthStatus: { status: 'logged_out', reason: 'no_card' },
+    },
+    {
+      cardStatus: { status: 'no_card' },
+      expectedAuthStatus: { status: 'logged_out', reason: 'no_card' },
+    },
+    {
+      cardStatus: { status: 'card_error' },
+      expectedAuthStatus: { status: 'logged_out', reason: 'card_error' },
+      expectedLog: [
+        LogEventId.AuthLogin,
+        'unknown',
+        {
+          disposition: LogDispositionStandardTypes.Failure,
+          message: 'User failed login.',
+          reason: 'card_error',
+        },
+      ],
+    },
+    {
+      cardStatus: { status: 'no_card' },
+      expectedAuthStatus: { status: 'logged_out', reason: 'no_card' },
+    },
+    {
+      cardStatus: { status: 'ready', user: undefined },
+      expectedAuthStatus: {
+        status: 'logged_out',
+        reason: 'invalid_user_on_card',
+      },
+      expectedLog: [
+        LogEventId.AuthLogin,
+        'unknown',
+        {
+          disposition: LogDispositionStandardTypes.Failure,
+          message: 'User failed login.',
+          reason: 'invalid_user_on_card',
+        },
+      ],
+    },
+    {
+      cardStatus: { status: 'no_card' },
+      expectedAuthStatus: { status: 'logged_out', reason: 'no_card' },
+    },
+    {
+      cardStatus: { status: 'ready', user: systemAdministratorUser },
+      expectedAuthStatus: {
+        status: 'checking_pin',
+        user: systemAdministratorUser,
+      },
+    },
+    {
+      cardStatus: { status: 'no_card' },
+      expectedAuthStatus: { status: 'logged_out', reason: 'no_card' },
+      expectedLog: [
+        LogEventId.AuthPinEntry,
+        'system_administrator',
+        {
+          disposition: LogDispositionStandardTypes.Failure,
+          message: 'User canceled PIN entry.',
+        },
+      ],
+    },
+  ];
+
+  let logIndex = 1;
+  for (const { cardStatus, expectedAuthStatus, expectedLog } of testSequence) {
+    mockCardStatus(cardStatus);
+    expect(await auth.getAuthStatus(machineState)).toEqual(expectedAuthStatus);
+    if (expectedLog) {
+      expect(mockLogger.log).toHaveBeenNthCalledWith(logIndex, ...expectedLog);
+      logIndex += 1;
+    }
+  }
+  expect(mockLogger.log).toHaveBeenCalledTimes(3);
+});
+
+test.each<{
+  description: string;
+  user: UserWithCard;
+}>([
+  {
+    description: 'system administrator',
+    user: systemAdministratorUser,
+  },
+  {
+    description: 'election manager',
+    user: electionManagerUser,
+  },
+])('Login and logout for users with PINs - $description', async ({ user }) => {
+  const auth = new InsertedSmartCardAuth({
+    card: mockCard,
+    config: defaultConfig,
+    logger: mockLogger,
+  });
+
+  mockCardStatus({ status: 'ready', user });
+  expect(await auth.getAuthStatus(machineState)).toEqual({
+    status: 'checking_pin',
+    user,
+  });
+
+  mockCard.checkPin
+    .expectCallWith(wrongPin)
+    .resolves({ response: 'incorrect', numRemainingAttempts: Infinity });
```

Comment:
Do we need any testing for what happens when there are no remaining pin attempts and the user tries to enter their pin? I can't remember what happens in that case and what code is responsible for it

Reply:
Yeah good q! I haven't actually implemented that functionality yet. Will be doing so as part of https://github.com/votingworks/vxsuite/issues/2527